### ISSUE TITLE
endian detection in sysdep.h is tightly depend on Linux and Windows

### ISSUE THREAD

**Issue by obache** [msgpack/msgpack-c#296]
in sysdep.h:
```
#include <arpa/inet.h>  /* __BYTE_ORDER */
#  if !defined(__APPLE__) && !defined(__FreeBSD__) && !defined(__OpenBSD__) && \
       !(defined(__sun) && defined(__SVR4))
#    include <byteswap.h>
#  endif
```
It is not good idea to add various "!defined(XXX)" cases.
 http://www.gnu.org/software/gnulib/manual/html_node/byteswap_002eh.html
If should be simply
```
#if defined(linux)
```
or
```
#if defined(__GLIBC__)
```

In the next block:
```
#if !defined(__LITTLE_ENDIAN__) && !defined(__BIG_ENDIAN__)
#if __BYTE_ORDER == __LITTLE_ENDIAN
#define __LITTLE_ENDIAN__
#elif __BYTE_ORDER == __BIG_ENDIAN
#define __BIG_ENDIAN__
#elif _WIN32
#define __LITTLE_ENDIAN__
#endif
#endif
```
It is not true that ```__BYTE_ORDER``` is always defined, for example, *BSD and Solaris.
For such case, ```__LITTLE_ENDIAN``` is also not defined, so ```__BYTE_ORDER == __LITTLE_ENDIAN``` is always TRUE, then misdetected as ```__LITTLE_ENDIAN__``` even if it is big endian.

Please refer  other products (for example X11/Xarch.h from xproto) how to detect machine byte order for such platforms.

**Comment by redboltz** [comment]
Thanks you for reporting the issue. 

## Checking byteswap.h

According to http://www.gnu.org/software/gnulib/manual/html_node/byteswap_002eh.html ,
> This header file is missing on some platforms: Mac OS X 10.5, FreeBSD 6.0, NetBSD 5.0, OpenBSD 3.8, Minix 3.1.8, AIX 5.1, HP-UX 11, IRIX 6.5, OSF/1 5.1, Solaris 11 2011-11, mingw, MSVC 9, Interix 3.5, BeOS.

> If should be simply
> #if defined(linux)
> or
> #if defined(__GLIBC__)

I think that ```__GLIBC__``` is defined in above platforms that don't have byteswap.h. So I need to use #if defined(linux). I'm not sure checking ```linux``` is enough, because other header files description indicate the different and smaller missing information as follows:
http://www.gnu.org/software/gnulib/manual/html_node/arpa_002finet_002eh.html#arpa_002finet_002eh

## Checking endian

I agree with you. I've checked X11/Xarch.h http://www.opensource.apple.com/source/X11proto/X11proto-40/xproto/xproto-7.0.15/Xarch.h?txt The license seems not to be common but it seems to be a permissive license like as Apache License version 2.0, the license of msgpack-c. If we refer to or copy it, what should we do? If I understand correctly, all we need to do is notice the original license and copyright holders in our sysdep.h. (I'm not 100% sure.)

I will also look for other endian checking code that is a common permissive license.

**Comment by obache** [comment]
''arpa/inet.h''' is a part of  standard headers, so it should be available for any modern POSIX system:
 http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/arpa_inet.h.html
Before XNS5, '"s" and "l" in ''ntohX''' may be really "short" and "long", and may not "uint16_t" and "unit32_t", but such ancient system is probably out of the scope of this product.
Main problem here is "missing standard 64bit way now".
One way is current "use byte swap macros (bswap_64(), bswap64() swap64(),...) for LITTLE ENDIAN",
Another way is "use ntohX() equivalent non standard 64bit ntohll(), or upcoming standard be64toh() family".

**Comment by redboltz** [comment]
Fixed by #312...
